newsletter/addressimport/parsers.py

```diff
@@ -0,0 +1,333 @@
+import logging
+logger = logging.getLogger(__name__)
+
+from django import forms
+from django.conf import settings
+from django.core.exceptions import ValidationError
+from django.core.validators import validate_email
+from django.utils.translation import ugettext_lazy as _, ugettext
+
+from newsletter.models import Subscription
+
+
+class AddressList(object):
+    """ List with unique addresses. """
+
+    def __init__(self, newsletter, ignore_errors=False):
+        self.newsletter = newsletter
+        self.ignore_errors = ignore_errors
+        self.addresses = {}
+
+    def add(self, email, name=None, location='unknown location'):
+        """ Add name to list. """
+
+        logger.debug("Going to add %s <%s>", name, email)
+
+        name = check_name(name, self.ignore_errors)
+        email = check_email(email, self.ignore_errors)
+
+        try:
+            validate_email(email)
+        except ValidationError:
+            logger.warn(
+                "Entry '%s' does not contain a valid e-mail address at %s."
+                % (email, location)
+            )
+
+            if not self.ignore_errors:
+                raise forms.ValidationError(_(
+                    "Entry '%s' does not contain a valid "
+                    "e-mail address.") % name
+                )
+
+        if email in self.addresses:
+            logger.warn(
+                "Entry '%s' contains a duplicate entry at %s."
+                % (email, location)
+            )
+
+            if not self.ignore_errors:
+                raise forms.ValidationError(_(
+                    "The address file contains duplicate entries "
+                    "for '%s'.") % email)
+
+        try:
+            subscription_exists(self.newsletter, email, name)
+        except ValidationError:
+            logger.warn(
+                "Entry '%s' is already subscribed to at %s."
+                % (email, location)
+            )
+            # "Entry '%s' at line %d is already subscribed to "
+            # "with email '%s'",
+            # name, myreader.line_num, email, extra=dict(data={'row': row}))
+
+            if not self.ignore_errors:
+                raise forms.ValidationError(
+                    _("Some entries are already subscribed to."))
+
+        self.addresses[email] = name
+
+
+def subscription_exists(newsletter, email, name=None):
+    """
+    Return wheter or not a subscription exists.
+    """
+    qs = Subscription.objects.filter(
+        newsletter__id=newsletter.id,
+        subscribed=True,
+        email_field__exact=email)
+
+    return not qs.exists()
+
+
+def check_email(email, ignore_errors=False):
+    """
+    Check (length of) email address.
+
+    TODO: Update this to perform full validation on email.
+    """
+
+    if settings.DEBUG:
+        logger.debug("Checking e-mail address %s", email)
+
+    email_length = \
+        Subscription._meta.get_field_by_name('email_field')[0].max_length
```

Comment:
In principle I totally agree with you guys. However, I really do like to work in split-screen and my eyesight does not allow to do this at very small font sizes. Besides this, spreading code over lines often aids in readability.

Perhaps we could agree on a 'hard' and a 'soft' limit - 119 being the 'hard' one and 79 the 'soft' one.